Add extension/theme landing pages

tofumatt opened the pull request:
Fixes #1285
Fixes #1286 

### Screenshots

![screen shot 2016-12-16 at 15 57 55](https://cloud.githubusercontent.com/assets/90871/21269038/045e86c6-c3a9-11e6-9d8f-1d8bbbada7cd.png)


![screen shot 2016-12-16 at 15 58 03](https://cloud.githubusercontent.com/assets/90871/21269043/08a69d54-c3a9-11e6-9465-5fbb97d54ad7.png)

![screen shot 2016-12-16 at 15 58 05](https://cloud.githubusercontent.com/assets/90871/21269052/0ca557e2-c3a9-11e6-96db-e5f2a4833c9d.png)

### Video

![dec-16-2016 16-03-42](https://cloud.githubusercontent.com/assets/90871/21269097/44fc71ca-c3a9-11e6-9827-282ba7d123c1.gif)

tofumatt commented:
r? Sorry I added the reviewer status but I'm guessing @kumar303 doesn't get the emails and I know they don't show up in assigned yet 😞 

I'll fix this conflict quickly in a moment.

kumar303 commented:
Argh, I totally do get the emails but they are no different than other people's review comments (which I typically skim over) so they are easy to mix. Yeah, I think assigning PRs to me is the best way because I check that from the dashboard regular. I'll dig into this...

tofumatt commented:
No probs, it’s easy to set both, I just didn’t want it to be too noisy.
Evidently I needed to be more noisy 😉 🥁

kumar303 commented on src/amo/utils.js:
```diff
@@ -0,0 +1,53 @@
+  const state = getState();
+  const addonType = params.addonType.replace(/s$/, '');
+
+  if (!(addonType === state.addonType && !state.loading)) {
```
Shouldn't this be `state.landing.addonType` and `state.lading.loading` ?

kumar303 commented on src/amo/utils.js:
```diff
@@ -0,0 +1,53 @@
+    return fetchLandingAddons({ addonType, api: state.api, dispatch });
+  }
+
+  return true;
```
I would prefer to see `Promise.resolve(true)` but maybe I'll give up. I hate `asyncConnect` so much.

tofumatt replied:
I actually agree, though @mstriemer pointed out there is an ever-so-slight performance difference in returning true.

But it's like `1ms` right? Do we care that much? I'm happy to use `Promise.resolve()` if we don't care, but I'm also obsessive enough that I'll take the odd API for a relatively easy `1ms` perf gain.

kumar303 replied:
one tick on the event loop? Maybe. Seems like a wasted optimization effort. What's nice about returning a promise is that we can use these functions as utilities in other functions more easily. It doesn't break `asyncConnect` either way so I don't feel strongly.

kumar303 commented on tests/client/amo/components/TestLandingPage.js:
```diff
@@ -0,0 +1,88 @@
+    return findDOMNode(findRenderedComponentWithType(renderIntoDocument(
+      <Provider store={createStore(initialState)}>
+        <I18nProvider i18n={getFakeI18nInst()}>
+          <LandingPage {...props} />
```
Because of all the HOC hotness, I think you can just do it like this now, right?

    <LandingPage store={createStore(initialState)} i18n={getFakeI18nInst()} {...props} />

tofumatt replied:
omg wow that's so much nicer, yeah 👍

tofumatt replied:
Actually no because sub-components need the `i18n` too. I could do shallow render though, I'll try that.

kumar303 commented on tests/client/amo/components/TestLandingPage.js:
```diff
@@ -0,0 +1,88 @@
+          name: 'Pop again',
+        },
+      ],
+    };
```
Did you try this by using real actions to set up the data? I think it would make the test better. It would look something like this:
````javascript
const store = createStore(initialState);
store.dispatch(setLanding({
  addonType: 'themes',
  featuredAddons: [
    {
      authors: [{}],
      name: 'Howdy again',
    },
  ],
  highlyRatedAddons: ...
}));
const root = render({ store });
````

tofumatt replied:
Agreed, not using them definitely led to an error earlier. I will move to that. Sorry I keep not doing it, I promise I'm learning, I think these tests are just a bit old.

tofumatt replied:
In this case I'm just testing the (dumb) component and involving actions seems to complicate the test and test things that are covered elsewhere in this test. This component doesn't even use the state so involving actions on this one seems too much.

tofumatt replied:
(I did change it to use `fakeAddon` though which tidied it up a bit.)

kumar303 replied:
Oh yeah, you're totally right. I thought this component was listening to state to get the data.

kumar303 commented on tests/client/amo/components/TestLandingPage.js:
```diff
@@ -0,0 +1,88 @@
+    const root = render(props);
+
+    assert.deepEqual(
+      Object.values(root.querySelectorAll('.SearchResult-heading'))
```
We generally use refs to get at dom things like this. I guess I don't feel too strongly about it but using refs would be more consistent.

tofumatt replied:
I often find the ref tests more annoying to write (and they clutter up the HTML), especially when I end up just wanted to test rendered output.

In this case I'm testing things rendered by subcomponents and I find this to be much easier without any harm than dependency injection and refs.

kumar303 replied:
Yeah, that makes sense. I use both.

kumar303 commented on tests/client/amo/containers/TestExtensionsPage.js:
```diff
@@ -0,0 +1,26 @@
+        highlyRated: { results: 'high' },
+        popular: { results: 'pop' },
+      },
+    };
```
Let me know if I'm off base here but I really think setting up raw state is dangerous. It makes it hard to iterate on the structure of our internal state. If you used actions to set up the state then it would mitigate that problem because you'd get test failures in all the right places. Is it too cumbersome to use the actions?

tofumatt replied:
No you're totally right. I'll fix it, we should never do this. 👍 👍 👍

kumar303 commented on tests/client/amo/test_utils.js:
```diff
@@ -0,0 +1,69 @@
+import { loadLandingAddons } from 'amo/utils';
+
+
+describe('AMO utils loadLandingAddons()', () => {
```
mocha output isn't as nice as node/py.test but when you're hunting for a test file, using real file paths is more helpful. I'd do it like this:
````javascript
describe('amo/utils', () => {
  describe('loadLandingAddons()', () => {
    ...
  });
});
````

tofumatt replied:
For sure. I'd love to tidy up all the test descriptions some lazy afternoon.

kumar303 commented on tests/client/amo/test_utils.js:
```diff
@@ -0,0 +1,69 @@
+        assert.strictEqual(loadLandingAddons({
+          store: {
+            dispatch: sinon.stub(),
+            getState: () => ({
```
this seems wrong. It should be returning `state.landing.addonType...` (I think?). Anyway, why do you need a stub store? Can't you just use the real one?

kumar303 commented on src/amo/containers/ExtensionsPage.js:
```diff
@@ -0,0 +1,27 @@
+
+
+export function setAddonType(state, ownProps) {
+  return { addonType: ownProps.params.addonType.replace(/s$/, '') };
```
We only use International English, not British English. (Just kidding.)

Ryuno-Ki replied:
Uhh, US English became international? I prefer BE. Closer to home :-)

tofumatt commented on src/amo/components/LandingPage.js:
```diff
@@ -0,0 +1,75 @@
+    const html = contentForTypes[addonType];
+
+    return (
+      <div className={`LandingPage LandingPage-${addonType}`}>
```
Quite, thanks for catching it. ES6 string interpolation makes me lazy 😉

tofumatt commented on src/amo/components/LandingPage.scss:
```diff
@@ -0,0 +1,9 @@
+.LandingPage {
+  padding: $padding-page;
+
+  .AddonsCard {
```
Totally right, thanks for noticing. I'll fix that up, should be easy.

tofumatt commented on src/amo/utils.js:
```diff
@@ -0,0 +1,53 @@
+  ];
+
+  return Promise.all(landingRequests)
+    .then((response) => dispatch(loadLanding({
```
Omg that works?! Cool.

tofumatt commented:
Style updated with blue small-caps:
![screen shot 2016-12-22 at 23 14 58](https://cloud.githubusercontent.com/assets/90871/21443358/7b81d7b0-c89c-11e6-82b1-fcd808291c0c.png)

tofumatt commented:
Okay, I addressed the PR comments I think and this is ready for another r?

I tweaked the style (screenshot above) and moved to actions for things pretty much everywhere which are much nicer and caught some bugs. 👍

kumar303 commented on src/amo/components/AddonsCard/index.js:
```diff
@@ -22,7 +22,8 @@ export default class AddonsCard extends React.Component {
         {addons && addons.length ? (
-          <ul className="AddonsCard-list">
+          <ul className="AddonsCard-list"
+            ref={(ref) => { this.cardList = ref; }}>
```
Hmm, is this ref used anywhere? I don't see it

kumar303 commented on src/amo/components/LandingPage.js:
```diff
@@ -0,0 +1,118 @@
+      extension: {
+        featuredHeader: i18n.gettext('Featured extensions'),
+        featuredFooter: (
+          <Link className="LandingPage-more-link" to="test">
```
`to="test"`? Is there another bug to fill these in?

kumar303 commented on src/amo/components/LandingPage.js:
```diff
@@ -0,0 +1,118 @@
+  contentForType(addonType) {
+    const { i18n } = this.props;
+
+    const contentForTypes = {
```
I see a lot of repetition here, especially with all the `<Link>` tags that mostly look the same. I think the only difference between `themes` and `extensions` will be the text and destinations. It would be nice to DRY this up with that in mind. It could happen in the next path when you fill in the destinations though.

kumar303 commented on src/amo/components/LandingPage.js:
```diff
@@ -0,0 +1,118 @@
+
+    return (
+      <div className={classNames('LandingPage', `LandingPage-${addonType}`)}>
+        <AddonsCard addons={featuredAddons} className="FeaturedAddons"
```
You might find it easy to DRY up the repetition of all the footer `<Link>`s by moving each of these into a new component:
````javascript
class FeaturedAddonCard extends React.Component {
  render() {
    const footer = (
      <Link className="LandingPage-more-link" to="test">{this.props.footerText}</Link>
    );
    return (
      <AddonsCard addons={this.props.highlyRatedAddons} className="{this.props.className}"
          header={this.props.headerText} footer={footer} />
    );
  }
}
````

tofumatt replied:
Yeah I thought that too and was gonna do it in a tidy-up bug to avoid too many components and more tests in this PR, but I'll do it here if you're cool with that.

kumar303 commented on src/amo/routes.js:
```diff
@@ -21,5 +22,6 @@ export default (
     <Route path=":addonType/:slug/" component={CategoryPage} />
     <Route path="fxa-authenticate" component={HandleLogin} />
     <Route path="search/" component={SearchPage} />
+    <Route path=":pluralAddonType/" component={LandingPage} />
```
thanks, the `pluralAddonType` makes it way less ambiguous now

kumar303 commented on src/amo/utils.js:
```diff
@@ -0,0 +1,32 @@
+    .then(([featured, highlyRated, popular]) => dispatch(
+      loadLanding({ addonType, featured, highlyRated, popular })
+    ))
+    .catch(() => dispatch(failLanding({ addonType })));
```
are failures displayed anywhere? I need to figure out how to make `asyncConnect` work with the new API error handler.

tofumatt replied:
Not yet, I wasn't sure how to hook them up, I just figured I should have an error action. Is it superfluous or is there a way I should do that?

kumar303 commented on src/amo/utils.js:
```diff
@@ -0,0 +1,32 @@
+  const state = getState();
+  const addonType = params.pluralAddonType.replace(/s$/, '');
+
+  if (!(addonType === state.landing.addonType && !state.landing.loading)) {
```
parsing `!!state.landing.loading` is hard on my brain. This would be more clear to me:
````javascript
if (addonType !== state.landing.addonType || state.landing.loading) {...}
````
Although now I'm confused about the loading part of that if statement.

tofumatt replied:
Yeah this is actually too much, modelled in my head off the more complex checks for search.

In this case `addonType !== state.landing.addonType` is fine though that would make the page eventually stale... though this data isn't nearly as changed as even search results.

Basically this shouldn't be here I think... I'll remove it.

kumar303 commented on tests/client/amo/components/TestLandingPage.js:
```diff
@@ -0,0 +1,117 @@
+  describe('mapStateToProps()', () => {
+    it('sets the props', () => {
+      const store = createStore({ application: 'android' });
+      // store.dispatch(landingActions.getLanding({ addonType }));
```
looks like an old comment

kumar303 commented on tests/client/amo/components/TestLandingPage.js:
```diff
@@ -0,0 +1,117 @@
+  });
+
+  it('renders each add-on when set', () => {
+    const props = {
```
I feel like this would be a great place to use `store.dispatch()` to populate the data because it will make sure that the rendering is tied directly to realistic state data. You can still still use `...fakeAddon` of course.

kumar303 commented on tests/client/amo/components/TestLandingPage.js:
```diff
@@ -0,0 +1,117 @@
+    }, 'No LandingPage content for addonType: XUL');
+  });
+
+  describe('mapStateToProps()', () => {
```
Oh, I get your comment about not deleting much code with this new patch. To be honest, I don't think we need this test as long as you use `store.dispatch()` in the rendering test above. 

We typically isolate state mapper tests from component tests but only for when a real `store` object would make testing the component too difficult. In this case, it doesn't.

To put it another way: if you remove this test you'll still get 100% code coverage because the other rendering test will have to rely on mapStateToProps() to pass.

tofumatt replied:
Removing this knocks test coverage down a bit but I'll move things around and see if I can use store for this, cheers. 👍

kumar303 commented on tests/client/amo/test_utils.js:
```diff
@@ -0,0 +1,68 @@
+    const addonType = 'theme';
+    let ownProps;
+
+    before(() => {
```
should this be `beforeEach` or do you only want it to run once?

tofumatt commented:
Alright, addressed everything and created an issue for the one bit that's outside the scope of this PR (and possibly mobile V1).

Good for another r?
